### PR TITLE
PublickKeyTokenProvider: Fix password update routine with password hash

### LINKED ISSUE #33757: Update password check not working as expected
Has anyone verified if this has any effect?

I just installed 23.0.8 and still see all the user's oc_authtokens being updated on basic auth requests.

My theory: this patch compares the saved encrypted password with the newly generated encrypted password. What it does not take into account is that openssl_public_encrypt outputs a different ciphertext every time, even on identical input values.

see https://stackoverflow.com/a/2980121 for an explanation.

_Originally posted by @Mask in https://github.com/nextcloud/server/issues/33485#issuecomment-1229540616_

### PR DESCRIPTION
**Pull request by marcelklehr**
fixes #33757

### CODE REVIEW

File: lib/private/Authentication/Token/PublicKeyTokenProvider.php [hunk 1]
```diff
@@ -270,10 +270,16 @@ public function setPassword(IToken $token, string $tokenId, string $password) {
...
 			$publicKey = $t->getPublicKey();
 			$t->setPassword($this->encryptPassword($password, $publicKey));
+			$t->setPasswordHash($this->hashPassword($password));
 			$this->updateToken($t);
 		}
 	}
 
+	private function hashPassword(string $password) : string {
+		$secret = $this->config->getSystemValue('secret');
+		return hash('sha512', $password . $secret);
```
Comment (CarlSchwan) [inline lib/private/Authentication/Token/PublicKeyTokenProvider.php:280]: should probably use password_verify https://www.php.net/manual/en/function.password-verify.php

Comment (marcelklehr) [reply lib/private/Authentication/Token/PublicKeyTokenProvider.php]: Good idea! Done

File: lib/private/Authentication/Token/PublicKeyTokenProvider.php [hunk 2]
```diff
@@ -270,10 +270,15 @@ public function setPassword(IToken $token, string $tokenId, string $password) {
 		foreach ($tokens as $t) {
 			$publicKey = $t->getPublicKey();
 			$t->setPassword($this->encryptPassword($password, $publicKey));
+			$t->setPasswordHash($this->hashPassword($password));
 			$this->updateToken($t);
 		}
 	}
 
+	private function hashPassword(string $password) : string {
+		return password_hash($password, PASSWORD_BCRYPT);
```
Comment (nickvergessen) [inline lib/private/Authentication/Token/PublicKeyTokenProvider.php:279]: should use https://github.com/nextcloud/server/blob/215aef3cbdc1963be1bb6bca5218ee0a4b7f1665/lib/private/Security/Hasher.php#L90 as discussed with `sha1($password) . $password` to make sure bcrypt changes if a password is longer than 72 chars and only changes at the end.

Comment (nickvergessen) [reply lib/private/Authentication/Token/PublicKeyTokenProvider.php:279]: Note that the result can be up to 255 chars so the db column needs adjusting

**Comment by marcelklehr** [comment]
Let's wait till this is merged @ldidry

**Comment by ldidry** [comment]
Ok, thx. :slightly_smiling_face:

**Comment by marcelklehr** [comment]
Did the changes, not sure why php7.4 fails

File: lib/private/Authentication/Token/PublicKeyTokenProvider.php [hunk 3]
```diff
@@ -270,10 +276,15 @@ public function setPassword(IToken $token, string $tokenId, string $password) {
 		foreach ($tokens as $t) {
 			$publicKey = $t->getPublicKey();
 			$t->setPassword($this->encryptPassword($password, $publicKey));
+			$t->setPasswordHash($this->hashPassword($password));
 			$this->updateToken($t);
 		}
 	}
 
+	private function hashPassword(string $password) : string {
```
Comment (nickvergessen) [inline lib/private/Authentication/Token/PublicKeyTokenProvider.php:284]:
```suggestion
	private function hashPassword(string $password): string {
```

**Comment by icewind1991** [comment]
follow up from https://github.com/nextcloud/server/pull/33485

File: core/Migrations/Version25000Date20220905140840.php [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+	 * @param IOutput $output
+	 * @param Closure $schemaClosure The `\Closure` returns a `ISchemaWrapper`
+	 * @param array $options
+	 * @return null|ISchemaWrapper
+	 */
+	public function changeSchema(IOutput $output, Closure $schemaClosure, array $options): ?ISchemaWrapper {
+		/** @var ISchemaWrapper $schema */
+		$schema = $schemaClosure();
+
+		$comments = $schema->getTable('authtoken');
```
Comment (ChristophWurst) [inline core/Migrations/Version25000Date20220905140840.php:47]:
```suggestion
		$authTokenTable = $schema->getTable('authtoken');
```

File: lib/private/Authentication/Token/PublicKeyToken.php [hunk 5]
```diff
@@ -98,6 +103,7 @@ public function __construct() {
 		$this->addType('uid', 'string');
 		$this->addType('loginName', 'string');
 		$this->addType('password', 'string');
+		$this->addType('passwordHash', 'string');
```
Comment (ChristophWurst) [inline lib/private/Authentication/Token/PublicKeyToken.php:106]: default is string, so this isn't necessary

Comment (marcelklehr) [reply lib/private/Authentication/Token/PublicKeyToken.php:106]: Mh. Why is it added for the other string values then? I think it makes sense to have it for completeness, even if it's not strictly necessary

Comment (ChristophWurst) [reply lib/private/Authentication/Token/PublicKeyToken.php:106]: it's fine to have it. it won't do any harm :)

File: core/Migrations/Version25000Date20220905140840.php (hunk 4 above)
Comment (ChristophWurst) [reply core/Migrations/Version25000Date20220905140840.php]: you have to also update the following lines

Comment (marcelklehr) [reply core/Migrations/Version25000Date20220905140840.php]: ah, shit :D

File: lib/private/Authentication/Token/PublicKeyTokenProvider.php [hunk 6]
```diff
@@ -60,18 +61,23 @@ class PublicKeyTokenProvider implements IProvider {
...
+
 	public function __construct(PublicKeyTokenMapper $mapper,
-								ICrypto $crypto,
-								IConfig $config,
-								LoggerInterface $logger,
-								ITimeFactory $time) {
+								ICrypto              $crypto,
+								IConfig              $config,
+								LoggerInterface      $logger,
+								ITimeFactory         $time,
```
Comment (juliushaertl) [inline lib/private/Authentication/Token/PublicKeyTokenProvider.php:71]: Maybe revert the additional spaces as the last line also doesn't have it ;)

File: lib/private/Authentication/Token/PublicKeyTokenProvider.php [hunk 7]
```diff
@@ -60,18 +61,23 @@ class PublicKeyTokenProvider implements IProvider {
 	/** @var CappedMemoryCache */
 	private $cache;
 
+	/** @var IHasher */
+	private $hasher;
```
Comment (juliushaertl) [inline lib/private/Authentication/Token/PublicKeyTokenProvider.php:65]:
```suggestion
	private IHasher $hasher;
```

File: lib/private/Authentication/Token/PublicKeyTokenProvider.php [hunk 8]
```diff
@@ -270,10 +276,15 @@ public function setPassword(IToken $token, string $tokenId, string $password) {
 		foreach ($tokens as $t) {
 			$publicKey = $t->getPublicKey();
 			$t->setPassword($this->encryptPassword($password, $publicKey));
+			$t->setPasswordHash($this->hashPassword($password));
 			$this->updateToken($t);
 		}
 	}
 
+	private function hashPassword(string $password): string {
+		return $this->hasher->hash(sha1($password) . $password);
```
Comment (juliushaertl) [inline lib/private/Authentication/Token/PublicKeyTokenProvider.php:285]: What's the reasoning behind using the concatenated sha1 + password?

Comment (marcelklehr) [reply lib/private/Authentication/Token/PublicKeyTokenProvider.php:284]: Depending on which algorithm is used in IHasher, only the first 70 chars are used for the hash (see https://www.php.net/manual/en/function.password-hash.php)

File: lib/private/Authentication/Token/PublicKeyTokenProvider.php (hunk 6 above)
Comment (marcelklehr) [reply lib/private/Authentication/Token/PublicKeyTokenProvider.php]: :+1:

File: apps/admin_audit/composer/autoload.php [hunk 9]
```diff
@@ -3,8 +3,21 @@
 // autoload.php @generated by Composer
 
 if (PHP_VERSION_ID < 50600) {
-    echo 'Composer 2.3.0 dropped support for autoloading on PHP <5.6 and you are running '.PHP_VERSION.', please upgrade PHP or use Composer 2.2 LTS via "composer self-update --2.2". Aborting.'.PHP_EOL;
-    exit(1);
+    if (!headers_sent()) {
```
Comment (juliushaertl) [inline apps/admin_audit/composer/autoload.php:6]: Please only commit the changed autoloader mappings

**Review by juliusknorr — APPROVED** [review]
Apart from the comment, please squash the fixups and check the lint failure. Otherwise good to get in from my side 👍

File: apps/admin_audit/composer/autoload.php (hunk 9 above)
Comment (marcelklehr) [reply apps/admin_audit/composer/autoload.php:6]: ah, I see

**Comment by juliushaertl** [comment]
Cypress failure looks related (500 error on the login)

**Review by artonge — COMMENTED** [review]
Maybe naive, but couldn't we compare the value of decrypted password ?

**Comment by juliushaertl** [comment]
> Maybe naive, but couldn't we compare the value of decrypted password ?

We cannot decrypt all tokens of a user to update them as we only got the user password when the update is called, not the actual cipher to decrypt the tokens (e.g. the app passwords).

**Review by artonge — APPROVED** [review]

**Comment by marcelklehr** [comment]
For some reason the migration isn't executed before the cypress tests run. Help is appreciated.

**Comment by artonge** [comment]
> For some reason the migration isn't executed before the cypress tests run. Help is appreciated.

I had the same issue in another PR. Let me see if I can come up with a generic solution

**Comment by artonge** [comment]
https://github.com/nextcloud/server/pull/35889 should fix the issue.

**Comment by marcelklehr** [comment]
Should be good to merge now

**Comment by juliushaertl** [comment]
Pushed another commit to bump the version and trigger the db upgrade on existing setups.

**Comment by juliushaertl** [comment]
Failure unrelated

File: lib/private/Authentication/Token/PublicKeyTokenProvider.php [hunk 10]
```diff
@@ -435,11 +446,12 @@ public function updatePasswords(string $uid, string $password) {
 
 		// Update the password for all tokens
 		$tokens = $this->mapper->getTokenByUser($uid);
+		$passwordHash = $this->hashPassword($password);
 		foreach ($tokens as $t) {
 			$publicKey = $t->getPublicKey();
-			$encryptedPassword = $this->encryptPassword($password, $publicKey);
-			if ($t->getPassword() !== $encryptedPassword) {
-				$t->setPassword($encryptedPassword);
+			if ($t->getPasswordHash() === null || $this->hasher->verify(sha1($password) . $password, $t->getPasswordHash())) {
```
Comment (nickvergessen) [inline lib/private/Authentication/Token/PublicKeyTokenProvider.php:452]: Did you maybe mean:
```suggestion
			if ($t->getPasswordHash() === null || !$this->hasher->verify(sha1($password) . $password, $t->getPasswordHash())) {
```

So the password is updated everytime it does NOT match the old stored password?

Comment (nickvergessen) [reply lib/private/Authentication/Token/PublicKeyTokenProvider.php:452]: But the actual problem is `$this->hasher->verify` takes like ~0.2 seconds

**Pull request merged**